### PR TITLE
Improve handling of custom types

### LINKED ISSUE #35205: Custom types can't be favorited
**Godot version:**
40f0649e5

**Issue description:**
Custom types can't be favorited in the node selection.
They also don't appear in the `Recent` section.

**Steps to reproduce:**
1. Install a plugin from the asset library (For example my `Replicator Node` plugin).
2. Open the `Add Node` dialog.
3. Favorite the custom type.

### PR DESCRIPTION
**Pull request by KoBeWi**
Originally started as a fix for #35205, but I got a bit carried away.

The PR improves how custom types (i.e. types added with `EditorPlugin.add_custom_type()`) are handled throughout editor:
- added some helper methods to check if custom type exists
- fixed a bug where custom types weren't properly saved in history by create dialog (fixes #35205)
- custom types will now display their icon in inspector categories
It looks bad with [#57286 ](https://github.com/godotengine/godot/issues/57286) though:
image
but that's another issue to be fixed separately.
- custom types are now recognized by script create dialog. However, since they are not a real types, the script path will be implicitly inserted instead

So about the "they are not a real types" part, I discovered that while going through code. Custom types are supposed to be a helper to add a node with a specific script. They also have a gray icon in scene tree dock, unlike `class_name` types:
image
(this should be probably made consistent)
They aren't supposed to be real classes. They don't even register their name anywhere, aside from some internal map in EditorData. I added a mention in the docs that clarifies it a bit.

Their advantage is that you can enable/disable a plugin and the type will be added/removed from editor, but none of your scripts will break, because the type is never referenced by name. Thanks to that they also don't clutter global scope with more class names.
The disadvantage is that you can't create them from script using their name. Also users probably don't expect them to work this way; I bet there are some open issues that are invalid for this reason.

Anyways, this rambling isn't really related to this PR. There were just voices that we should consider removing custom types in 4.0, so the above text could be a good reference for a proposal.

### CODE REVIEW

File: editor/editor_inspector.cpp [hunk 1]
```diff
@@ -2543,10 +2543,13 @@ void EditorInspector::update_tree() {
...
 					StringName name = EditorNode::get_editor_data().script_class_get_name(script->get_path());
 					String icon_path = EditorNode::get_editor_data().script_class_get_icon_path(name);
-					if (name != StringName() && icon_path.length()) {
+					if (name != StringName() && !icon_path.is_empty()) {
 						category->icon = ResourceLoader::load(icon_path, "Texture");
 						break;
 					}
+					if (const EditorData::CustomType *ctype = EditorNode::get_editor_data().get_custom_type_by_path(script->get_path())) {
+						category->icon = ctype->icon;
+					}
```
Comment (YuriSizov) [inline editor/editor_inspector.cpp:2552]: Does this fall under the forbidden craft of "assignment in conditions"? Guess that may need to be expanded into multiple lines.

Also, shouldn't it break on match?

**Comment by YuriSizov** [comment]
I think this looks fine, besides a small note above. We didn't settle on removing this, so I guess we aren't. Didn't test it, would appreciate a rebase.

Also maybe @willnationsdev can give it a check? It touches on some stuff related to his work.

**Review by YuriSizov — APPROVED** [review]
Code looks good, and the mentioned features work.

One minor problem I've noticed is that the icon for the node, when creating a scene from a custom type, is not displayed correctly initially, but fixes itself after the scene is reloaded.

**Comment by akien-mga** [comment]
Thanks!

**Pull request merged**